portal/src/main/scala/cromwell/pipeline/service/ProjectService.scala

```diff
@@ -12,13 +13,22 @@ class ProjectService(projectRepository: ProjectRepository)(implicit executionCon
   def getProjectById(projectId: ProjectId): Future[Option[Project]] =
     projectRepository.getProjectById(projectId)
 
-  def addProject(request: ProjectAdditionRequest): Future[ProjectId] = {
+  def getProjectByName(namePattern: String, userId: UserId): Future[Option[Project]] = {
+    val result = projectRepository.getProjectByName(namePattern)
+    result.flatMap {
+      case Some(project) if project.ownerId == userId => result
+      case Some(_)                                    => Future.failed(new IllegalArgumentException("Access denied"))
+      case None                                       => Future.failed(new IllegalArgumentException("Project does not exist"))
```

Comment:
You have to use _differrent_ types of exceptions in situations where access is denied and when the project is not found. In this case, you can distinguish between them and assign them the appropriate error codes.

Reply:
Today I created custom types for this exceptions.